### PR TITLE
[SYCL][Graph] Add support for enabling CommandBuffer submission profiling

### PR DESCRIPTION
**Pull request by mfrancepillois**
Adds the support for enabling the profiling of CommandBuffer execution.
- Sycl event impl object contains a flag indicating that this event results from a graph submission. This flag was not set for every type of graph submission. This commit fixes this bug.
- Adds a extra event in the first command list associated to the CommandBuffer execution to obtain the start time of the graph execution.
- Modifies the urEventGetProfilingInfo function to get the CommandBuffer start time from this new event.
- Adds a hack to overload the submit time if the estimated time shift is too large.
- Adds an e2e test that checks the profiling info from an event returned by a graph submission.
- Removes the test checking for exception throwing (unsupported feature).
- Updates design doc

Closes Issue: #96

### CODE REVIEW

File: sycl/test-e2e/Graph/event_profiling_info.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    // Run graphs
+    event CopyEvent = Queue.submit(
+        [&](handler &CGH) { CGH.ext_oneapi_graph(CopyGraphExec); });
+    event KernelEvent = Queue.submit(
+        [&](handler &CGH) { CGH.ext_oneapi_graph(KernelGraphExec); });
+
+    Queue.wait_and_throw();
+
+    // Checks profiling times
+    assert(verifyProfiling(CopyEvent) && verifyProfiling(KernelEvent));
```
Comment (EwanC) [inline sycl/test-e2e/Graph/event_profiling_info.cpp]: Can we also test submitting the same graph twice, and verifying that the timestamps are different.

e.g 
```cpp
event KerneEvent1 = Queue.submit(
        [&](handler &CGH) { CGH.ext_oneapi_graph(KernelGraphExec); });
event KerneEvent2 = Queue.submit(
        [&](handler &CGH) { CGH.ext_oneapi_graph(KernelGraphExec); });

// Check that submit/start/end timestamps aren't identical for KernelEvent1 and KernelEvent2

```

**Review by EwanC — COMMENTED** [review]
LGTM, but a suggestion for improving test further

File: sycl/test-e2e/Graph/event_profiling_info.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+// from graph submission with event::get_profiling_info().
+// It first tests a graph made exclusively of memory operations,
+// then tests a graph made of kernels.
+// The second run is to check that there are no leaks reported with the embedded
+// ZE_DEBUG=4 testing capability.
+
+#include "./graph_common.hpp"
+#include <unistd.h>
+
+// #define VERBOSE
```
Comment (EwanC) [inline sycl/test-e2e/Graph/event_profiling_info.cpp]: Rather than commenting out the define, preference for
```cpp
#define VERBOSE 0

#if VERBOSE
// prints
#endif
```

File: sycl/plugins/unified_runtime/ur/adapters/level_zero/command_buffer.cpp [hunk 3]
```diff
@@ -814,6 +815,22 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
...
+    // the profiling engine to retrieve it.
+    UR_CALL(createEventAndAssociateQueue(Queue, &StartEvent,
+                                         UR_COMMAND_COMMAND_BUFFER_ENQUEUE_EXP,
+                                         WaitCommandList, false));
+
+    ZE2UR_CALL(zeCommandListAppendBarrier,
+               (WaitCommandList->first, StartEvent->ZeEvent, 1,
+                &(CommandBuffer->WaitEvent->ZeEvent)));
+
+    RetEvent->CommandData = StartEvent;
```
Comment (EwanC) [inline sycl/plugins/unified_runtime/ur/adapters/level_zero/command_buffer.cpp]: Do avoid doing this work all the time, can we only do it if the queue has `UR_QUEUE_FLAG_PROFILING_ENABLE` set?

Might also be worth adding something to the design doc about this approach https://github.com/reble/llvm/blob/sycl-graph-develop/sycl/doc/design/CommandGraph.md#level-zero (will be a faff to update the diagram so can just put in some text)

File: sycl/plugins/unified_runtime/ur/adapters/level_zero/command_buffer.cpp [hunk 4]
```diff
@@ -801,6 +801,7 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
 
   // Execution event for this enqueue of the PI command-buffer
   ur_event_handle_t RetEvent{};
+  ur_event_handle_t StartEvent{};
```
Comment (EwanC) [inline sycl/plugins/unified_runtime/ur/adapters/level_zero/command_buffer.cpp]: Could maybe move this definition into the scope on line 825 underneath the comment

**Review by EwanC — APPROVED** [review]
This approach LGTM

File: sycl/plugins/unified_runtime/ur/adapters/level_zero/command_buffer.cpp (hunk 3 above)
Comment (mfrancepillois) [reply sycl/plugins/unified_runtime/ur/adapters/level_zero/command_buffer.cpp]: I added the condition and updated the design doc

File: sycl/plugins/unified_runtime/ur/adapters/level_zero/command_buffer.cpp [hunk 5]
```diff
@@ -816,6 +816,25 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
     ZE2UR_CALL(zeCommandListAppendBarrier,
                (SignalCommandList->first, RetEvent->ZeEvent, 1,
                 &(CommandBuffer->SignalEvent->ZeEvent)));
+
+    if ((Queue->Properties & UR_QUEUE_FLAG_PROFILING_ENABLE)) {
+      // We create an extra-signal specific to the current execution of the
```
Comment (julianmi) [inline sycl/plugins/unified_runtime/ur/adapters/level_zero/command_buffer.cpp]:
```suggestion
      // We create an additional signal specific to the current execution of the
```

File: sycl/plugins/unified_runtime/ur/adapters/level_zero/command_buffer.cpp [hunk 6]
```diff
@@ -816,6 +816,25 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
...
+      // CommandBuffer. This signal is needed for profiling the execution time
+      // of the CommandBuffer. It waits for the WaitEvent to be signaled
+      // which indicates the start of the CommandBuffer actual execution.
+      // This event is embeded into the Event return to user to allow
```
Comment (julianmi) [inline sycl/plugins/unified_runtime/ur/adapters/level_zero/command_buffer.cpp]:
```suggestion
      // This event is embeded into the Event return to the user to allow
```

File: sycl/doc/design/CommandGraph.md [hunk 7]
```diff
@@ -166,7 +166,12 @@ created on UR command-buffer enqueue.
 There is also a *WaitEvent* used by the `ur_exp_command_buffer_handle_t` class
 in the prefix to wait on any dependencies passed in the enqueue wait-list.
 
-![L0 command-buffer diagram](images/L0_UR_command-buffer.svg)
+If command-buffer is about to be submitted to a queue with the profiling
```
Comment (julianmi) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
If a command-buffer is about to be submitted to a queue with the profiling 
```

File: sycl/source/detail/event_impl.cpp [hunk 8]
```diff
@@ -278,17 +278,31 @@ void event_impl::checkProfilingPreconditions() const {
...
+  // a estimated clock and not on the real device clock, may be ahead of time
+  // than the start time, which is based on the actual device clock.
+  // MSubmitTime is set in a critical performance path.
+  // Force reading the device clock when setting MSubmitTime may deteriorate
+  // the performance.
+  // Since submit time is a estimated time, we implement this little hack
+  // that allows all profiled time to be meaningful.
+  // (Note that the observed time deviation between the estimated clock and
+  // the real device clock  is less than 0.5ms. The approximation we make by
+  // forcing the re-sync of submt time to start time is less than 0.5ms).
```
Comment (julianmi) [inline sycl/source/detail/event_impl.cpp]:
```suggestion
  // forcing the re-sync of submit time to start time is less than 0.5ms).
```

File: sycl/source/detail/event_impl.cpp [hunk 9]
```diff
@@ -278,17 +278,31 @@ void event_impl::checkProfilingPreconditions() const {
...
+  // can be short. As a consequence, the submission time, which is based on
+  // a estimated clock and not on the real device clock, may be ahead of time
+  // than the start time, which is based on the actual device clock.
+  // MSubmitTime is set in a critical performance path.
+  // Force reading the device clock when setting MSubmitTime may deteriorate
+  // the performance.
+  // Since submit time is a estimated time, we implement this little hack
+  // that allows all profiled time to be meaningful.
+  // (Note that the observed time deviation between the estimated clock and
+  // the real device clock  is less than 0.5ms. The approximation we make by
```
Comment (julianmi) [inline sycl/source/detail/event_impl.cpp]: Where does this 0.5ms come from?

File: sycl/source/detail/graph_impl.cpp [hunk 10]
```diff
@@ -662,6 +661,7 @@ exec_graph_impl::enqueue(const std::shared_ptr<sycl::detail::queue_impl> &Queue,
       NewEvent = sycl::detail::Scheduler::getInstance().addCG(
           std::move(CommandGroup), Queue);
     }
+    NewEvent->setEventFromSubmitedExecCommandBuffer(true);
```
Comment (julianmi) [inline sycl/source/detail/graph_impl.cpp]: This is not part of this PR but should we fix this typo to `setEventFromSubmittedExecCommandBuffer` since you are touching this? Same for `isEventFromSubmitedExecCommandBuffer`.

File: sycl/unittests/Extensions/CommandGraph.cpp [hunk 11]
```diff
@@ -1685,76 +1685,6 @@ TEST_F(CommandGraphTest, BindlessExceptionCheck) {
   sycl::free(ImgMemUSM, Ctxt);
 }
 
-TEST_F(CommandGraphTest, GetProfilingInfoExceptionCheck) {
```
Comment (julianmi) [inline sycl/unittests/Extensions/CommandGraph.cpp:1688]: Does it make sense to rephrase that into a unittest checking that the events are set and freed correctly?

**Review by julianmi — COMMENTED** [review]
Do we have comparing data with wallclock times and potentially pure SYCL profiling data to estimate how accurate this approach is?

File: sycl/source/detail/event_impl.cpp (hunk 9 above)
Comment (mfrancepillois) [reply sycl/source/detail/event_impl.cpp]: It is empirical value (not super reliable but just to give an idea). On my setup the time shift was approx. 400us. On Ben setup the time shift was approx. 200us.

File: sycl/source/detail/graph_impl.cpp (hunk 10 above)
Comment (mfrancepillois) [reply sycl/source/detail/graph_impl.cpp]: I corrected this typo

**Comment by mfrancepillois** [comment]
> Do we have comparing data with wallclock times and potentially pure SYCL profiling data to estimate how accurate this approach is?

I added some timing info in the verbose version of the test to help verify the time given by the profiling against the wallclock time. On my setup, the time difference for submissions with approximate submit time is less than 100us (during which Sycl functions are crossed to go from user code to the actual location where task timestamps are taken).

File: sycl/unittests/Extensions/CommandGraph.cpp (hunk 11 above)
Comment (mfrancepillois) [reply sycl/unittests/Extensions/CommandGraph.cpp:1688]: I'm not sure I understand what you would like to test.
The `event_status_querying.cpp` test already verifies that events states are correctly set.

File: sycl/source/detail/event_impl.cpp (hunk 9 above)
Comment (julianmi) [reply sycl/source/detail/event_impl.cpp]: Thanks. Would you extend the comment with it being empirical please?

File: sycl/unittests/Extensions/CommandGraph.cpp (hunk 11 above)
Comment (julianmi) [reply sycl/unittests/Extensions/CommandGraph.cpp:1688]: I was wondering if we can test internally with an unittest whether the start event is correctly set, used (e.g., if it is taken from the first command-list), and freed. I think this could make sense since we extend the typical profiling path.

Comment (mfrancepillois) [reply sycl/unittests/Extensions/CommandGraph.cpp:1688]: This type of test could be relevant but unitests do not use actual backends but only a plug-in mock. So we can't test that in an unitest.
And (assuming we can re-read a level-zero commandlist, which I'm not sure we can do) I don't see this type of tests consistent with the purpose of e2e tests.

File: sycl/source/detail/event_impl.cpp (hunk 9 above)
Comment (mfrancepillois) [reply sycl/source/detail/event_impl.cpp]: I enhanced the comment this way.

File: sycl/unittests/Extensions/CommandGraph.cpp (hunk 11 above)
Comment (julianmi) [reply sycl/unittests/Extensions/CommandGraph.cpp:1688]: True. I missed that this is only L0. Please disregard my comments then.

**Review by julianmi — APPROVED** [review]

File: sycl/doc/design/CommandGraph.md [hunk 12]
```diff
@@ -166,7 +166,12 @@ created on UR command-buffer enqueue.
 There is also a *WaitEvent* used by the `ur_exp_command_buffer_handle_t` class
 in the prefix to wait on any dependencies passed in the enqueue wait-list.
 
-![L0 command-buffer diagram](images/L0_UR_command-buffer.svg)
+If a command-buffer is about to be submitted to a queue with the profiling 
+property enabled, a profiling *StartEvent* is added to the end of the prefix.
+This event is used to get the command-buffer execution start time on the 
+device.
```
Comment (Bensuo) [inline sycl/doc/design/CommandGraph.md:172]: Forgive me if this has already been discussed elsewhere, is there a reason we can't add the start event to the beginning of the actual graph command list on command buffer creation? I'd guess because we need access to the queue to associate the event, but maybe I'm wrong.

We're introducing some inaccuracy here I would think , although I don't know how big or how static that extra time is, and maybe there's no way around it.

File: sycl/test-e2e/Graph/event_profiling_info.cpp [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+// This test checks the profiling of an event returned
+// from graph submission with event::get_profiling_info().
+// It first tests a graph made exclusively of memory operations,
+// then tests a graph made of kernels.
+// The second run is to check that there are no leaks reported with the embedded
+// ZE_DEBUG=4 testing capability.
+
+#include "./graph_common.hpp"
+#include <unistd.h>
```
Comment (Bensuo) [inline sycl/test-e2e/Graph/event_profiling_info.cpp]: What are we using this header for here?

**Review by Bensuo — APPROVED** [review]

File: sycl/doc/design/CommandGraph.md (hunk 12 above)
Comment (mfrancepillois) [reply sycl/doc/design/CommandGraph.md:172]: I agree on the accuracy loss, but we need an unique event for a submission.
Indeed, since a command-buffer can be submitted multiple times, if we rely on the same event for each submission (i.e. an event associated with the command-buffer) the previous profiling info will be override by subsequent submissions.
And, we cannot add a submission unique event to the actual graph command list since this list is closed during the finalize step.

I would add that the accuracy issue is even double, since the end time is also taken for an event which belongs to the suffix command-list which may be not executed as soon as the actual graph command-list is complete.

File: sycl/plugins/unified_runtime/ur/adapters/level_zero/command_buffer.cpp [hunk 14]
```diff
@@ -816,6 +816,25 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
     ZE2UR_CALL(zeCommandListAppendBarrier,
                (SignalCommandList->first, RetEvent->ZeEvent, 1,
                 &(CommandBuffer->SignalEvent->ZeEvent)));
+
+    if ((Queue->Properties & UR_QUEUE_FLAG_PROFILING_ENABLE)) {
+      // We create an additional signal specific to the current execution of the
+      // CommandBuffer. This signal is needed for profiling the execution time
+      // of the CommandBuffer. It waits for the WaitEvent to be signaled
+      // which indicates the start of the CommandBuffer actual execution.
+      // This event is embeded into the Event return to the user to allow
```
Comment (reble) [inline sycl/plugins/unified_runtime/ur/adapters/level_zero/command_buffer.cpp]:
```suggestion
      // This event is embedded into the Event return to the user to allow
```

File: sycl/source/detail/event_impl.cpp [hunk 15]
```diff
@@ -278,17 +278,33 @@ void event_impl::checkProfilingPreconditions() const {
...
-  }
 }
 
 template <>
 uint64_t
 event_impl::get_profiling_info<info::event_profiling::command_submit>() {
   checkProfilingPreconditions();
+  // The delay between the submission and the actual start of a CommandBuffer
+  // can be short. As a consequence, the submission time, which is based on
+  // a estimated clock and not on the real device clock, may be ahead of time
```
Comment (reble) [inline sycl/source/detail/event_impl.cpp]:
```suggestion
  // can be short. Consequently, the submission time, which is based on
  // an estimated clock and not on the real device clock, may be ahead of time
```

File: sycl/source/detail/event_impl.cpp [hunk 16]
```diff
@@ -278,17 +278,33 @@ void event_impl::checkProfilingPreconditions() const {
...
+  // than the start time, which is based on the actual device clock.
+  // MSubmitTime is set in a critical performance path.
+  // Force reading the device clock when setting MSubmitTime may deteriorate
+  // the performance.
+  // Since submit time is a estimated time, we implement this little hack
```
Comment (reble) [inline sycl/source/detail/event_impl.cpp]:
```suggestion
  // Since submit time is an estimated time, we implement this little hack
```

File: sycl/source/detail/event_impl.cpp [hunk 17]
```diff
@@ -278,17 +278,33 @@ void event_impl::checkProfilingPreconditions() const {
...
+  // that allows all profiled time to be meaningful.
+  // (Note that the observed time deviation between the estimated clock and
+  // the real device clock  is less than 0.5ms. The approximation we make by
```
Comment (reble) [inline sycl/source/detail/event_impl.cpp]:
```suggestion
  // the real device clock is typically less than 0.5ms. The approximation we make by
```

File: sycl/test-e2e/Graph/event_profiling_info.cpp [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+
+#define VERBOSE 0
+
+#if VERBOSE
```
Comment (reble) [inline sycl/test-e2e/Graph/event_profiling_info.cpp]: We might want to rename this macro to make it more descriptive. E.g. `VERBOSE_PRINT` or even `GRAPH_TESTS_VERBOSE_PRINT`. Also do we need a numerical value here? or can we just set it to undefined by default like:

```suggestion
//#define GRAPH_TESTS_VERBOSE_PRINT

#ifdef GRAPH_TESTS_VERBOSE_PRINT
```

**Review by reble — APPROVED** [review]
LGTM!

**Comment by reble** [comment]
Let's create an upstream PR with profiling related changes only.

File: sycl/test-e2e/Graph/event_profiling_info.cpp (hunk 13 above)
Comment (mfrancepillois) [reply sycl/test-e2e/Graph/event_profiling_info.cpp]: You are right. I removed it.

File: sycl/test-e2e/Graph/event_profiling_info.cpp (hunk 18 above)
Comment (mfrancepillois) [reply sycl/test-e2e/Graph/event_profiling_info.cpp]: I changed the name for a more explicit name : GRAPH_TESTS_VERBOSE_PRINT.
However, I kept the numerical value to avoid having commented lines in our test (according to Ewan's comment previously).
